### PR TITLE
feature/SOF-4000

### PR DESCRIPTION
**Pull request by mohammadidinani**

### CODE REVIEW

File: express/parsers/utils.py [hunk 1]
```diff
@@ -16,3 +16,47 @@ def find_file(name, path):
...
+        else:
+            element_counts.append({
+                "count": 1,
+                "value": element["value"]
+            })
+        previous_element = basis["elements"][index]
+    return element_counts
+
+
+def to_poscar(lattice, basis):
```
Comment (timurbazhirov) [inline express/parsers/utils.py]: 1. lattic_basis_to_poscar
2. either append "_cartesian" to the function name of add a parameter with units to handle crystal coordinates too

File: express/parsers/apps/espresso/formats/txt.py [hunk 2]
```diff
@@ -285,6 +285,23 @@ def initial_lattice_vectors(self, text):
             lattice["vectors"][key] = [e * lattice_alat * Constant.BOHR for e in lattice["vectors"][key]]
         return lattice
 
+    def _extract_basis(self, text, lattice_alat, number_of_atoms):
```
Comment (timurbazhirov) [inline express/parsers/apps/espresso/formats/txt.py]: _extract_basis_in_angstroms

File: express/parsers/apps/espresso/formats/txt.py [hunk 3]
```diff
@@ -370,7 +379,7 @@ def _extract_data_from_bfgs_blocks(self, text, func):
             results.append(func(block))
         return results
 
-    def _extract_lattice(self, text, last_value=False, regex="lattice"):
+    def _extract_lattice(self, text, regex="lattice"):
```
Comment (timurbazhirov) [inline express/parsers/apps/espresso/formats/txt.py]: Same as above - which units is the lattice extracted in - angstrom?

Comment (mohammadidinani) [reply express/parsers/apps/espresso/formats/txt.py]: Done.

File: express/parsers/utils.py (hunk 1 above)
Comment (mohammadidinani) [reply express/parsers/utils.py]: Done.

File: express/parsers/apps/espresso/formats/txt.py [hunk 4]
```diff
@@ -275,19 +278,43 @@ def initial_lattice_vectors(self, text):
                     'a': [-0.561154473, -0.000000000, 0.561154473],
                     'b': [-0.000000000, 0.561154473, 0.561154473],
                     'c': [-0.561154473, 0.561154473, 0.000000000],
-                    'alat': 9.44858082
+                    'alat': 1
                 }
             }
         """
+        lattice_alat = self._lattice_alat(text)
```
Comment (timurbazhirov) [inline express/parsers/apps/espresso/formats/txt.py]: 1. Rename `self._lattice_alat` => `self._get_alat` in the class
2. `lattice_alat` => `alat` in the function
3. On line 286:
`lattice_in_alat_units` = self._extract_lattice(text, regex="lattice_alat")

File: express/parsers/apps/espresso/formats/txt.py [hunk 5]
```diff
@@ -275,19 +278,43 @@ def initial_lattice_vectors(self, text):
...
+
+    def _number_of_atoms(self, text):
+        return self._general_output_parser(text, **settings.REGEX["number_of_atoms"])[0]
+
     def initial_basis(self, text):
         """
-        Extracts initial basis from a given text. The initial basis is in alat format and needs to be converted.
+        Extracts initial basis from a given text.
+
+        Note: The initial basis is in alat format and hence it needs to be converted to angstrom.
```
Comment (timurbazhirov) [inline express/parsers/apps/espresso/formats/txt.py]: `Units: cartesian, angstrom` would suffice

File: express/parsers/apps/espresso/formats/txt.py [hunk 6]
```diff
@@ -307,21 +334,16 @@ def initial_basis(self, text):
...
-        matches = self._general_output_parser(text, **settings.REGEX["basis_alat"](number_of_atoms))
-        for idx, match in enumerate(matches):
-            basis["elements"].append({"id": idx, "value": match[0]})
-            coordinate = [float(match[1]), float(match[2]), float(match[3])]
-            coordinate = map(lambda x: x * lattice_alat * Constant.BOHR, coordinate)  # alat to angstrom
-            basis["coordinates"].append({"id": idx, "value": coordinate})
-
+        lattice_alat = self._lattice_alat(text)
+        number_of_atoms = self._number_of_atoms(text)
+        basis = self._extract_basis(text[text.find("positions (alat units)"):], number_of_atoms)
```
Comment (timurbazhirov) [inline express/parsers/apps/espresso/formats/txt.py]: `basis_in_alat_units = ...

File: express/parsers/utils.py [hunk 7]
```diff
@@ -16,3 +16,47 @@ def find_file(name, path):
         for file in files:
             if name in file:
                 return os.path.join(root, file)
+
+
+def find_files(name, path):
```
Comment (timurbazhirov) [inline express/parsers/utils.py]: find_fines_by_name_substring(substring, path)

**Pull request merged**